### PR TITLE
Fix #40174: Azure.Extensions.AspNetCore.DataProtection.Keys Heavy load causes blocked threads

### LINKED ISSUE #40174: [BUG] Azure.Extensions.AspNetCore.DataProtection.Keys Heavy load causes blocked threads
### Library name and version

Azure.Extensions.AspNetCore.DataProtection.Keys 1.2.2

### Describe the bug

We are using this together with ASP.NET Cookie authentication. One of our applications receives very many requests. When we restarted this application, it stopped responding. We took a memory dump, and it showed that all threads were waiting on 

https://github.com/Azure/azure-sdk-for-net/blob/9af180c0f0c5e31ba087c8209a01e68283df719e/sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureKeyVaultXmlDecryptor.cs#L27

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#avoid-using-taskresult-and-taskwait

### Expected behavior

It should handle heavy load during startup.

### Actual behavior

All threads were blocked.

### Reproduction Steps

NA

### Environment

.NET 7

### PR DESCRIPTION
**Pull request by AlexanderSher**
From the description of the issue, it looks like a case of ThreadPool starvation, which is reproducible in test environment. There are multiple task-related issues in Azure.Extensions, so this PR provides the following changes with the goal to prevent ThreadPool starvation:
- Changes `ParallelSecretLoader.AddSecretToLoad` to explicitly create a task to schedule `SecretClient.GetSecretAsync` instead of waiting for the task returned by underlying code
- `ParallelSecretLoader.WaitForAll` now uses `Task.WaitAll` and is synchronous (asynchronous version is called `WaitForAllAsync()`
- `AzureKeyVaultConfigurationProvider.Load` has separate implementation which uses synchronous version of `SecretClient` methods and `ParallelSecretLoader.WaitForAll`.
- `AzureKeyVaultXmlEncryptor` and `AzureKeyVaultXmlDecryptor` use synchronous version of the client methods and don't block-wait the ThreadPool thread on the newly created task.

There are two known open questions about `ClientRegistration<TClient>` type implementation:
- `ClientRegistration.GetClient` calls `_factory` delegate under lock
- `ClientRegistration.Dispose`, when client implements `IDisposableAsync` and doesn't implement `IDisposable`, calls `DisposeAsync` and synchronously waits on it.

Moving `_factory` call out of the lock or removing `DisposeAsync` call from `ClientRegistration.Dispose` would result in semantical changes, so I'm leaving those outside of this PR.

### CODE REVIEW

**Comment by jsquire** [comment]
> ClientRegistration.Dispose, when client implements IDisposableAsync and doesn't implement IDisposable, calls DisposeAsync and synchronously waits on it.

**This should stay as-is, unless you've got an alternate suggestion for how to address the corner case. **

This is intended to handle the scenario when registered components are not following guidance and are running on an older version of the framework.   Modern versions of .NET (net6.0+) will call `DisposeAsync` on any registered component that implements it, and `Dispose` on those that do not.  Legacy versions call only `Dispose` and do not have awareness of the async interface.  Components should implement the sync version as well as async, though that is not always the case.

Because our infrastructure can be used to register components that we do not own nor control, we deemed it less risky use sync-over-async to ensure cleanup in this specific corner case than to bypass cleanup entirely and hope that there's a finalizer to take care of it.

File: sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs [hunk 1]
```diff
@@ -71,63 +63,87 @@ public TClient GetClient(IServiceProvider serviceProvider, object options, Token
...
+                _cachedClient = default;
+            }
 
-                    disposableClient.Dispose();
-                }
-                else if (_asyncDisposable)
-                {
-                    DisposeAsync().GetAwaiter().GetResult();
-                }
+            if (disposableClient is not null)
```
Comment (JoshLove-msft) [inline sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs:140]: Should asyncDisposableClient take precedence?

Comment (JoshLove-msft) [reply sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs:140]: Ah I guess it makes sense since this is the Dispose call.

**Review by JoshLove-msft — APPROVED** [review]
LGTM but do we have any before/after perf numbers or memory dumps comparing threadpool usage?

File: sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs (hunk 1 above)
Comment (jsquire) [reply sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs:140]: Yeah, the original implementation intentionally preferred the sync version for `Dispose` and the async version for `DisposeAsync` and fell back only when there's no corresponding version.

**Review by heaths — APPROVED** [review]
KV changes look good, and I agree with @JoshLove-msft about getting metrics. Should be equivalent, but it would be good to confirm.

**Review by jsquire — APPROVED** [review]

**Pull request merged**